### PR TITLE
Booking integrated into main!

### PR DESCRIPTION
**Pull request by Apulatjonov**

### CODE REVIEW

File: src/main/java/com/exadel/coolDesking/booking/BookingService.java [hunk 1]
```diff
@@ -1,10 +1,105 @@
...
+
+    @Autowired
+    public BookingService(BookingRepository bookingRepository, UserRepository userRepository, WorkplaceRepository workplaceRepository, WaitingListService waitingListService) {
+        this.bookingRepository = bookingRepository;
+        this.userRepository = userRepository;
+        this.workplaceRepository = workplaceRepository;
+    }
+
+    @Bean
+    private ModelMapper modelMapper() {
```
Comment (viktarkolbik) [inline src/main/java/com/exadel/coolDesking/booking/BookingService.java:36]: 1. ModelMapper often has specific configuration for different domains, so reusing one instance in many situations can cause issues.
2. Why do you create a Bean, and then just call method _modelMapper()_? You newer reuse this bean so it's useless.
3. Do mapping work in separate class.

File: src/main/java/com/exadel/coolDesking/booking/BookingService.java [hunk 2]
```diff
@@ -1,10 +1,105 @@
...
+    public List<Booking> getBookings() {
+        return bookingRepository.findAll();
+    }
+
+    public BookingResponseDTO addBooking(BookingDTO bookingDto) {
+        Booking booking = modelMapper().map(bookingDto, Booking.class);
+//        if (checkBooking(bookingDto)) {
+//            return new BookingResponseDTO(409, "Conflict", "The selected workplace has been booked by another user");
+//        }
+        if (userRepository.findById(bookingDto.getUserId()).get().getRole().equals(UserRole.COMMON_USER)) {
```
Comment (viktarkolbik) [inline src/main/java/com/exadel/coolDesking/booking/BookingService.java:51]: Don't check access manually. 
@gargatuma is busy with Security, ask him for help

File: src/main/java/com/exadel/coolDesking/booking/BookingService.java [hunk 3]
```diff
@@ -1,10 +1,105 @@
...
+            return new BookingResponseDTO(403, "Forbidden", "You don’t have permissions to book this workplace");
+        }
+        Optional<User> user = userRepository.findById(bookingDto.getUserId());
+        booking.setWorkplace(workplaceRepository.findById(bookingDto.getWorkplaceId()).get());
+        booking.setUser(user.get());
+        bookingRepository.save(booking);
+        return new BookingResponseDTO(201, "Created", null);
```
Comment (viktarkolbik) [inline src/main/java/com/exadel/coolDesking/booking/BookingService.java:58]: Here and in other places
**201** and **"Created"** are useless. You must return it as status code of the Response.

File: src/main/java/com/exadel/coolDesking/booking/Booking.java [hunk 4]
```diff
@@ -27,32 +27,32 @@
...
     private User user;
 
-    @NotNull
+//    @NotNull
     @ManyToOne
     @JoinColumn(name = "workplace_id", referencedColumnName = "id")
     private Workplace workplace;
 
-    @NotBlank
+//    @NotNull
```
Comment (viktarkolbik) [inline src/main/java/com/exadel/coolDesking/booking/Booking.java:40]: Drop useless comments.

File: src/main/resources/application.properties [hunk 5]
```diff
@@ -1,7 +1,7 @@
 spring.datasource.url=jdbc:postgresql://localhost:5432/hotdesking
 spring.datasource.driver-class-name=org.postgresql.Driver
-spring.datasource.username=aweasadbek
-spring.datasource.password=1234
+spring.datasource.username=postgres
```
Comment (viktarkolbik) [inline src/main/resources/application.properties:3]: don't change credentials

File: src/main/java/com/exadel/coolDesking/booking/BookingService.java [hunk 6]
```diff
@@ -1,10 +1,105 @@
...
+        Optional<Workplace> workplace = workplaceRepository.findById(bookingDto.getWorkplaceId());
+        if (workplace.isEmpty() || user.isEmpty()) {
+            return new BookingResponseDTO(409, "Conflict", "Your booking can't be updated by selected parameters");
+        }
+        Booking booking = modelMapper().map(bookingDto, Booking.class);
+
+        booking.setWorkplace(workplace.get());
+        booking.setUser(user.get());
+        bookingRepository.save(booking);
+        return new BookingResponseDTO(200, "OK", "Your booking has been successfully updated");
```
Comment (viktarkolbik) [inline src/main/java/com/exadel/coolDesking/booking/BookingService.java:81]: Convert updated entity into DTO and return it

File: src/main/java/com/exadel/coolDesking/booking/BookingService.java [hunk 7]
```diff
@@ -1,10 +1,105 @@
...
+        return modelMapper;
+    }
+
+    public List<Booking> getBookings() {
+        return bookingRepository.findAll();
+    }
+
+    public BookingResponseDTO addBooking(BookingDTO bookingDto) {
+        Booking booking = modelMapper().map(bookingDto, Booking.class);
+//        if (checkBooking(bookingDto)) {
```
Comment (viktarkolbik) [inline src/main/java/com/exadel/coolDesking/booking/BookingService.java:48]: Drop useless comment

**Comment by Apulatjonov** [comment]
Everything mentioned above has been fixed!

File: src/main/resources/application.properties [hunk 8]
```diff
@@ -1,7 +1,7 @@
 spring.datasource.url=jdbc:postgresql://localhost:5432/hotdesking
 spring.datasource.driver-class-name=org.postgresql.Driver
-spring.datasource.username=postgres
-spring.datasource.password=Pulat7343
+spring.datasource.username=aweasadbek
```
Comment (Apulatjonov) [inline src/main/resources/application.properties:3]: I need to change it for testing purpose

File: src/main/java/com/exadel/coolDesking/booking/BookingController.java [hunk 9]
```diff
@@ -1,24 +1,42 @@
...
+    }
+
+    @PostMapping
+    public ResponseEntity<?> addBooking(@RequestBody BookingDTO bookingDTO) {
+        return ResponseEntity.status(HttpStatus.CREATED).body(bookingService.addBooking(bookingDTO));
+    }
+
+    @PutMapping("/{id}")
+    public ResponseEntity<?> updateBooking(@PathVariable UUID id, @RequestBody BookingDTO bookingDto) {
+        return ResponseEntity.status(HttpStatus.CREATED).body(bookingService.updateBooking(bookingDto, id));
```
Comment (viktarkolbik) [inline src/main/java/com/exadel/coolDesking/booking/BookingController.java:34]: For update operation return "OK" status

File: src/main/java/com/exadel/coolDesking/booking/BookingRepository.java [hunk 10]
```diff
@@ -3,8 +3,11 @@
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.stereotype.Repository;
 
+import java.util.Date;
```
Comment (viktarkolbik) [inline src/main/java/com/exadel/coolDesking/booking/BookingRepository.java:6]: java.util.Date is outdated. Use LocalDate

File: pom.xml [hunk 11]
```diff
@@ -79,7 +79,12 @@
             <artifactId>poi-ooxml</artifactId>
             <version>5.2.0</version>
         </dependency>
-    </dependencies>
```
Comment (viktarkolbik) [inline pom.xml:82]: Why do you need this dependency?

File: src/main/java/com/exadel/coolDesking/booking/BookingResponseDTO.java [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+import lombok.Setter;
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Getter
+@Setter
+public class BookingResponseDTO {
+    private Integer code;
+    private String status;
+    private Object message;
```
Comment (viktarkolbik) [inline src/main/java/com/exadel/coolDesking/booking/BookingResponseDTO.java:15]: Why object? Why not String?

File: src/main/java/com/exadel/coolDesking/booking/BookingResponseDTO.java [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Getter
+@Setter
+public class BookingResponseDTO {
+    private Integer code;
```
Comment (viktarkolbik) [inline src/main/java/com/exadel/coolDesking/booking/BookingResponseDTO.java:13]: code and status are useless here

File: pom.xml (hunk 11 above)
Comment (Apulatjonov) [reply pom.xml:82]: I tried to fix the error I was getting in workplaceService when pulled the project from main branch that time

File: src/main/java/com/exadel/coolDesking/booking/BookingController.java (hunk 9 above)
Comment (Apulatjonov) [reply src/main/java/com/exadel/coolDesking/booking/BookingController.java:34]: got you

File: src/main/java/com/exadel/coolDesking/booking/BookingResponseDTO.java (hunk 13 above)
Comment (Apulatjonov) [reply src/main/java/com/exadel/coolDesking/booking/BookingResponseDTO.java:13]: ok

File: src/main/java/com/exadel/coolDesking/booking/BookingResponseDTO.java (hunk 12 above)
Comment (Apulatjonov) [reply src/main/java/com/exadel/coolDesking/booking/BookingResponseDTO.java:15]: I was giving "objects" itself not "objects.toString()", ok I will fix it

File: src/main/java/com/exadel/coolDesking/booking/Booking.java [hunk 14]
```diff
@@ -27,32 +25,26 @@
...
-    @Column(name = "start")
-    private LocalDate start;
+    @Column(name = "start_date")
+    private Date startDate;
 
-    @NotNull
-    @Column(name = "end")
-    private LocalDate end;
+    @Column(name = "end_date")
+    private Date endDate;
```
Comment (gargatuma) [inline src/main/java/com/exadel/coolDesking/booking/Booking.java:49]: add validation

**Review by gargatuma — CHANGES_REQUESTED** [review]

File: src/main/java/com/exadel/coolDesking/booking/BookingController.java [hunk 15]
```diff
@@ -1,24 +1,42 @@
...
 public class BookingController {
+
     private final BookingService bookingService;
 
+    @GetMapping
+    public ResponseEntity<?> getAll() {
+        return ResponseEntity.ok(bookingService.getBookings());
+    }
+
     @GetMapping(path = "/{id}")
```
Comment (gargatuma) [inline src/main/java/com/exadel/coolDesking/booking/BookingController.java:22]: no need for path

Comment (gargatuma) [reply src/main/java/com/exadel/coolDesking/booking/BookingController.java:22]: just "/{id}"

File: src/main/java/com/exadel/coolDesking/booking/BookingService.java [hunk 16]
```diff
@@ -1,10 +1,92 @@
...
+        }
+        Booking booking = bookingMapper.bookingCreateDtoToEntity(bookingDto);
+        Booking savedBooking = bookingRepository.save(booking);
+        return bookingMapper.entityToResponseDTO(savedBooking);
+    }
+
+
+    public BookingDTO getBooking(UUID bookingId) {
+        Booking booking = bookingRepository.findById(bookingId)
+                .orElseThrow(() -> new NotFoundException("There is no such booking", Booking.class, "bookingId"));
```
Comment (gargatuma) [inline src/main/java/com/exadel/coolDesking/booking/BookingService.java:63]: for field parameter pass just "id" not "bookingId"

File: src/main/java/com/exadel/coolDesking/booking/BookingService.java [hunk 17]
```diff
@@ -1,10 +1,92 @@
...
+        return bookingMapper.entityToResponseDTO(booking);
+    }
+
+    public BookingDTO updateBooking(BookingDTO bookingDto, UUID bookingID) {
+        Booking booking = bookingRepository.findById(bookingID)
+                .orElseThrow(() -> new NotFoundException("There is no such booking", Booking.class, "bookingId"));
```
Comment (gargatuma) [inline src/main/java/com/exadel/coolDesking/booking/BookingService.java:69]: for field parameter pass just "id" not "bookingId"

File: src/main/java/com/exadel/coolDesking/booking/BookingService.java [hunk 18]
```diff
@@ -1,10 +1,92 @@
...
+
+        bookingMapper.updateEntity(bookingDto, booking);
+        Booking savedBookings = bookingRepository.save(booking);
+        return bookingMapper.entityToResponseDTO(savedBookings);
+    }
+
+    public Boolean deleteBooking(UUID bookingId) {
+        Booking booking = bookingRepository.findById(bookingId)
+                .orElseThrow(() -> new NotFoundException("There is no such booking", Booking.class, "bookingId"));
```
Comment (gargatuma) [inline src/main/java/com/exadel/coolDesking/booking/BookingService.java:78]: for field parameter pass just "id" not "bookingId"

File: src/main/java/com/exadel/coolDesking/booking/BookingService.java [hunk 19]
```diff
@@ -1,10 +1,92 @@
...
+
+        bookingRepository.delete(booking);
+        return Boolean.TRUE;
```
Comment (gargatuma) [inline src/main/java/com/exadel/coolDesking/booking/BookingService.java:81]: you can return just "true" and java autoboxing wil work

File: src/main/java/com/exadel/coolDesking/booking/BookingDTO.java [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+@Setter
+public class BookingDTO {
+    private UUID workplaceId;
+    private UUID userId;
+    private Date startDate;
+    private Date endDate;
+    private Boolean isRecurring;
+    private Boolean hasParking;
+    private Integer frequency;
+}
```
Comment (gargatuma) [inline src/main/java/com/exadel/coolDesking/booking/BookingDTO.java:19]: missing validations

File: src/main/java/com/exadel/coolDesking/booking/BookingController.java [hunk 21]
```diff
@@ -1,24 +1,42 @@
...
-    public ResponseEntity<Booking> getById(@PathVariable UUID id) {
-        return null;
+    public ResponseEntity<?> getById(@PathVariable UUID id) {
+        return ResponseEntity.ok(bookingService.getBooking(id));
+    }
+
+    @PostMapping
+    public ResponseEntity<?> addBooking(@RequestBody BookingDTO bookingDTO) {
```
Comment (gargatuma) [inline src/main/java/com/exadel/coolDesking/booking/BookingController.java:28]: add validation

File: src/main/java/com/exadel/coolDesking/booking/BookingController.java [hunk 22]
```diff
@@ -1,24 +1,42 @@
...
+        return ResponseEntity.status(HttpStatus.CREATED).body(bookingService.addBooking(bookingDTO));
+    }
+
+    @PutMapping("/{id}")
+    public ResponseEntity<?> updateBooking(@PathVariable UUID id, @RequestBody BookingDTO bookingDto) {
```
Comment (gargatuma) [inline src/main/java/com/exadel/coolDesking/booking/BookingController.java:33]: add validation

File: src/main/java/com/exadel/coolDesking/booking/BookingService.java [hunk 23]
```diff
@@ -1,10 +1,92 @@
...
+        bookings.forEach(booking -> {
+            BookingDTO bookingDTO = bookingMapper.entityToResponseDTO(booking);
+            bookingDtos.add(bookingDTO);
+        });
+        return bookingDtos;
+    }
+
+    public BookingDTO addBooking(BookingDTO bookingDto) {
+        if (checkBooking(bookingDto)) {
+            throw new ConflictException("Conflict", Booking.class, "booking");
```
Comment (gargatuma) [inline src/main/java/com/exadel/coolDesking/booking/BookingService.java:53]: for field parameter pass just "id" not "bookingId"

**Review by gargatuma — CHANGES_REQUESTED** [review]

File: src/main/resources/db/changelog/changeset/10-drop-wrong-end-keyword-booking-table.yaml [hunk 24]
```diff
@@ -0,0 +1,17 @@
+databaseChangeLog:
+  - changeSet:
+      id: drop-end-column
+      author: Apulatjonov
+      preConditions:
+        - onFail: MARK_RAN
+          -columnExists:
+            tableName: booking
+            columnName: end
```
[the remark this answers is not in the document]
Comment (Apulatjonov) [reply src/main/resources/db/changelog/changeset/10-drop-wrong-end-keyword-booking-table.yaml:9]: I did

File: src/main/java/com/exadel/coolDesking/booking/BookingController.java (hunk 15 above)
Comment (Apulatjonov) [reply src/main/java/com/exadel/coolDesking/booking/BookingController.java:22]: it is autogenerated

**Review by viktarkolbik — DISMISSED** [review]

File: pom.xml (hunk 11 above)
Comment (viktarkolbik) [reply pom.xml:82]: Still don't understand what type of the error this dependency fixes.

File: src/main/java/com/exadel/coolDesking/booking/BookingService.java [hunk 25]
```diff
@@ -1,10 +1,82 @@
...
+        this.bookingRepository = bookingRepository;
+        this.userRepository = userRepository;
+        this.workplaceRepository = workplaceRepository;
+        this.bookingMapper = bookingMapper;
+    }
+
+    public List<BookingDTO> getBookings() {
+        List<BookingDTO> bookingDtos = new ArrayList<>();
+        List<Booking> bookings = bookingRepository.findAll();
+        bookings.forEach(booking -> {
```
Comment (gargatuma) [inline src/main/java/com/exadel/coolDesking/booking/BookingService.java:34]: u can use stream map here

**Review by gargatuma — CHANGES_REQUESTED** [review]

File: src/main/java/com/exadel/coolDesking/booking/BookingService.java (hunk 23 above)
Comment (Apulatjonov) [reply src/main/java/com/exadel/coolDesking/booking/BookingService.java:43]: could not get?!

File: src/main/java/com/exadel/coolDesking/booking/BookingDTO.java [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+@AllArgsConstructor
+@NoArgsConstructor
+@Data
+public class BookingDTO {
+
+    private UUID id;
+
+    private UUID workplaceId;
+
+    private UUID userId;
```
[the remark this answers is not in the document]
Comment (Apulatjonov) [reply src/main/java/com/exadel/coolDesking/booking/BookingDTO.java:21]: ok

File: src/main/java/com/exadel/coolDesking/booking/BookingService.java [hunk 27]
```diff
@@ -1,10 +1,83 @@
...
 @RequiredArgsConstructor
 @Service
 public class BookingService {
     private final BookingRepository bookingRepository;
+    private final UserRepository userRepository;
+    private final WorkplaceRepository workplaceRepository;
+    private final BookingMapper bookingMapper;
+
+    @Autowired
+    public BookingService(BookingRepository bookingRepository, UserRepository userRepository, WorkplaceRepository workplaceRepository, WaitingListService waitingListService, BookingMapper bookingMapper) {
```
[the remark this answers is not in the document]
Comment (Apulatjonov) [reply src/main/java/com/exadel/coolDesking/booking/BookingService.java:24]: ok

File: src/main/java/com/exadel/coolDesking/booking/BookingService.java [hunk 28]
```diff
@@ -1,10 +1,83 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
 @RequiredArgsConstructor
 @Service
 public class BookingService {
     private final BookingRepository bookingRepository;
+    private final UserRepository userRepository;
+    private final WorkplaceRepository workplaceRepository;
```
[the remark this answers is not in the document]
Comment (Apulatjonov) [reply src/main/java/com/exadel/coolDesking/booking/BookingService.java:20]: ok

**Review by viktarkolbik — APPROVED** [review]

File: src/main/java/com/exadel/coolDesking/booking/BookingController.java (hunk 15 above)
Comment (viktarkolbik) [reply src/main/java/com/exadel/coolDesking/booking/BookingController.java:23]: simplify your code
```suggestion
    @GetMapping("/{id}")/
```

**Review by gargatuma — APPROVED** [review]

**Pull request merged**